### PR TITLE
[editor] Implement fields validation in editor.

### PR DESCRIPTION
**Pull request by mgsergio**
@ygorshenin @Zverik

### CODE REVIEW

File: indexer/editable_map_object.cpp [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#include "search/v2/postcodes_matcher.hpp"
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: indexer should not depend on search.

Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: This will break linux build.

File: indexer/editable_map_object.cpp [hunk 2]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+  // TODO: Improve this basic limit - it was choosen by @Zverik.
+  auto constexpr kMaxHouseNumberLength = 15;
+  if (us.size() > kMaxHouseNumberLength)
+    return false;
+
+  // TODO: Should we allow arabic numbers like U+0661 ١	Arabic-Indic Digit One?
+  strings::NormalizeDigits(us);
+  for (strings::UniChar const c : us)
+  {
+    // Valid house numbers contain at least one number.
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: s/number/digit/

File: indexer/editable_map_object.cpp [hunk 3]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+    if (c >= '0' && c <= '9')
+      return true;
+  }
+  return false;
+}
+
+// static
+bool EditableMapObject::ValidateFlats(string const & flats)
+{
+  auto flatRangeIterator = strings::SimpleTokenizer(flats, ",;");
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: s/flatRangeIterator/it/g

File: indexer/editable_map_object.cpp [hunk 4]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+  for (; flatRangeIterator != strings::SimpleTokenizer(); ++flatRangeIterator)
+  {
+    auto flatRange = *flatRangeIterator;
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: s/flatRange/range/

File: indexer/editable_map_object.cpp [hunk 5]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+    strings::Trim(flatRange);
+    vector<string> range(strings::SimpleTokenizer(flatRange, "-"), strings::SimpleTokenizer());
+    if (range.size() > 2)
+      return false;
+
+    for (auto const & rangeBorder : range)
+    {
+      if (!all_of(begin(rangeBorder), end(rangeBorder), isalnum))
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: isalnum for utf8? Seriously?

File: indexer/editable_map_object.cpp [hunk 6]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+  return search::v2::LooksLikePostcode(postCode, false /* check whole matching */);
+}
+
+// static
+bool EditableMapObject::ValidatePhone(string const & phone)
+{
+  auto start = begin(phone);
+  auto const stop = end(phone);
+
+  auto const startsWithPlus = *start == '+';
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: Please, add a test where ValidatePhone is called with an empty string.

File: indexer/editable_map_object.cpp [hunk 7]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+
+  auto const maxNumberLen = 12;
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: s/maxNumberLen/kMaxNumberLen/
s/minNumberLen/kMinNumberLen/

File: indexer/editable_map_object.cpp [hunk 8]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+
+  if (startsWithPlus)
+    ++start;
+
+  auto digitsCount = 0;
+  for (; start != stop; ++start)
+  {
+    auto const isCharValid = isdigit(*start) || *start == '(' ||
+                             *start == ')' || *start == ' ' || *start == '-';
+    if (isdigit(*start))
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: Please, check isCharValid first and then increment digitsCount if isdigit(*start).

File: indexer/editable_map_object.cpp [hunk 9]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+}
+
+// static
+bool EditableMapObject::ValidateWebsite(string const & site)
+{
+  if (site.empty())
+    return true;
+
+  auto const dotPos = find(begin(site), end(site), '.');
+  // site should contain at least one dot but not at the begining and no at the and.
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: Sentences must start with a capital letter.

File: indexer/editable_map_object.cpp [hunk 10]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+bool EditableMapObject::ValidateEmail(string const & email)
+{
+  if (email.empty())
+    return true;
+
+  auto const atPos = find(begin(email), end(email), '@');
+  if (atPos == end(email))
+    return false;
+
+  // There should be only one '@' sign.
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: http://stackoverflow.com/questions/12355858/how-many-symbol-can-be-in-an-email-address

**Comment by mgsergio** [comment]
DUDE JTTP

File: indexer/editable_map_object.cpp (hunk 3 above)
Comment (Zverik) [inline indexer/editable_map_object.cpp]: Только `;`, запятая — это ошибка.

Comment (Zverik) [inline indexer/editable_map_object.cpp]: Можно сначала заменить все запятые на `;`, кстати.

File: indexer/editable_map_object.cpp (hunk 5 above)
Comment (Zverik) [inline indexer/editable_map_object.cpp]: why not? `all_of(..., isalnum)` looks sane.

File: indexer/editable_map_object.cpp [hunk 11]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+{
+  auto start = begin(phone);
+  auto const stop = end(phone);
+
+  auto const startsWithPlus = *start == '+';
+
+  auto const maxNumberLen = 12;
+  auto const minNumberLen = 7;
+
+  if (startsWithPlus)
```
Comment (Zverik) [inline indexer/editable_map_object.cpp]: Эта переменная больше нигде не используется — может, не объявлять её, а прямо здесь `*start == '+'`?

File: indexer/editable_map_object.cpp [hunk 12]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+// static
+bool EditableMapObject::ValidatePhone(string const & phone)
+{
+  auto start = begin(phone);
+  auto const stop = end(phone);
+
+  auto const startsWithPlus = *start == '+';
+
+  auto const maxNumberLen = 12;
+  auto const minNumberLen = 7;
```
Comment (Zverik) [inline indexer/editable_map_object.cpp]: Реально 5 (в глубинке это может быть `2-34-57`).

File: indexer/editable_map_object.cpp (hunk 7 above)
Comment (Zverik) [inline indexer/editable_map_object.cpp]: 15 (https://en.wikipedia.org/wiki/E.164)

File: indexer/editable_map_object.cpp (hunk 9 above)
Comment (Zverik) [inline indexer/editable_map_object.cpp]: `...and not the the end.`

File: indexer/editable_map_object.cpp (hunk 10 above)
Comment (Zverik) [inline indexer/editable_map_object.cpp]: Если у какого-то владельца магазина `@` в первой части почтового адреса, то он ССЗБ.

File: indexer/indexer_tests/editable_map_object_test.cpp [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+using osm::EditableMapObject;
+
+UNIT_TEST(EditableMapObject_ValidateBuildingLevels)
+{
+  TEST(EditableMapObject::ValidateBuildingLevels("13"), ());
+  TEST(EditableMapObject::ValidateBuildingLevels("17"), ());
+  TEST(EditableMapObject::ValidateBuildingLevels("25"), ());
+  TEST(!EditableMapObject::ValidateBuildingLevels("26"), ());
+  TEST(!EditableMapObject::ValidateBuildingLevels("ab"), ());
+}
```
Comment (Zverik) [inline indexer/indexer_tests/editable_map_object_test.cpp]: +пустую строку
+одна цифра (вместо 17, например)
+40 цифр (переполнение при преобразовании)

File: indexer/indexer_tests/editable_map_object_test.cpp [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+  TEST(EditableMapObject::ValidateFlats("123a"), ());
+  TEST(EditableMapObject::ValidateFlats("a"), ());
+  TEST(EditableMapObject::ValidateFlats("123-456,a-e"), ());
+  TEST(EditableMapObject::ValidateFlats("123-456"), ());
+  TEST(EditableMapObject::ValidateFlats("123-456;43-45"), ());
+  TEST(EditableMapObject::ValidateFlats("123-456,43-45"), ());
+  TEST(EditableMapObject::ValidateFlats("123,456, 43,45"), ());
+  TEST(!EditableMapObject::ValidateFlats("234-234 124"), ());
+  TEST(!EditableMapObject::ValidateFlats("123-345-567"), ());
+  TEST(!EditableMapObject::ValidateFlats("234-234;234("), ());
```
Comment (Zverik) [inline indexer/indexer_tests/editable_map_object_test.cpp]: `-;`

File: indexer/editable_map_object.cpp (hunk 10 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: то он что/кто?

Comment (mgsergio) [inline indexer/editable_map_object.cpp]: АА, буратино)

File: indexer/indexer_tests/editable_map_object_test.cpp [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+UNIT_TEST(EditableMapObject_ValidateWebsite)
+{
+  TEST(EditableMapObject::ValidateWebsite(""), ());
+  TEST(EditableMapObject::ValidateWebsite("qwe.rty"), ());
+
+  TEST(!EditableMapObject::ValidateWebsite("qwerty"), ());
+  TEST(!EditableMapObject::ValidateWebsite(".qwerty"), ());
+  TEST(!EditableMapObject::ValidateWebsite("qwerty."), ());
+  TEST(!EditableMapObject::ValidateWebsite(".qwerty."), ());
```
Comment (Zverik) [inline indexer/indexer_tests/editable_map_object_test.cpp]: так а что с `http://`?

File: indexer/editable_map_object.cpp [hunk 16]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+{
+  return ValidatePhone(fax);
+}
+
+// static
+bool EditableMapObject::ValidateWebsite(string const & site)
+{
+  if (site.empty())
+    return true;
+
```
Comment (Zverik) [inline indexer/editable_map_object.cpp]: Где-то нужно автодобавлять `http://`, и в валидаторе проверять, что сайт начинается либо с `http://`, либо с `https://`.

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 14 above)
Comment (mgsergio) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Что это значит?

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 15 above)
Comment (mgsergio) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Нет смысла проверять, если все равно надо подставлять.
По крайней мере не в валидации проверку делать, а в сеттере.

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 14 above)
Comment (Zverik) [inline indexer/indexer_tests/editable_map_object_test.cpp]: добавь на проверку эту строчку

File: indexer/editable_map_object.cpp (hunk 5 above)
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: Does alnum works fine for cyrillic letters?

**Comment by Zverik** [comment]
Результат тестирования: http://angola.mapsme.mail.msk:8080/job/GitHubWatchdog/1078/

File: indexer/editable_map_object.cpp (hunk 3 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Почему запятая ошибка?
В примерах полно такого.

File: indexer/editable_map_object.cpp [hunk 17]
```diff
@@ -213,4 +185,134 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+  strings::UniString us = strings::MakeUniString(houseNumber);
+  // TODO: Improve this basic limit - it was choosen by @Zverik.
+  auto constexpr kMaxHouseNumberLength = 15;
+  if (us.size() > kMaxHouseNumberLength)
+    return false;
+
+  // TODO: Should we allow arabic numbers like U+0661 ١	Arabic-Indic Digit One?
+  strings::NormalizeDigits(us);
+  for (strings::UniChar const c : us)
```
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 2 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 3 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: s/flatRangeIterator/it/g — Done.

File: indexer/editable_map_object.cpp (hunk 4 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 6 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 7 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: All Done.

File: indexer/editable_map_object.cpp (hunk 8 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 3 above)
Comment (Zverik) [inline indexer/editable_map_object.cpp]: Запятая удобнее при вводе, но её [нет в спецификации](http://wiki.openstreetmap.org/wiki/Key:addr:flats). Серии значений в осме разделяются точками с запятой.

File: indexer/editable_map_object.cpp (hunk 5 above)
Comment (Zverik) [inline indexer/editable_map_object.cpp]: fair point.

File: indexer/editable_map_object.cpp (hunk 9 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 11 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 12 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 13 above)
Comment (mgsergio) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Done.

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 14 above)
Comment (mgsergio) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 5 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: @ygorshenin, what is your proposal?

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 15 above)
Comment (mgsergio) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 16 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 3 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 1 above)
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: For those who don't believe: 
* add a unit test in indexer/indexer_tests/editable_map_object_test.cpp for EditableMapObject::ValidatePostcode()
* try to build it with static linkage
* receive
```
clang++ -ccc-gcc-name g++ -stdlib=libc++ -lprofiler -Wl,--gc-sections -Wl,-Bsymbolic-functions -Wl,-rpath,/home/y/Qt/5.5/gcc_64 -Wl,-rpath,/home/y/Qt/5.5/gcc_64/lib -o ../out/debug/MAPS.ME ../out/debug/tmp/qt/about.o ../out/debug/tmp/qt/create_feature_dialog.o ../out/debug/tmp/qt/draw_widget.o ../out/debug/tmp/qt/editor_dialog.o ../out/debug/tmp/qt/info_dialog.o ../out/debug/tmp/qt/main.o ../out/debug/tmp/qt/mainwindow.o ../out/debug/tmp/qt/osm_auth_dialog.o ../out/debug/tmp/qt/place_page_dialog.o ../out/debug/tmp/qt/preferences_dialog.o ../out/debug/tmp/qt/proxystyle.o ../out/debug/tmp/qt/qtoglcontext.o ../out/debug/tmp/qt/qtoglcontextfactory.o ../out/debug/tmp/qt/search_panel.o ../out/debug/tmp/qt/slider_ctrl.o ../out/debug/tmp/qt/update_dialog.o ../out/debug/tmp/qt/qrc_resources.o ../out/debug/tmp/qt/moc_about.o ../out/debug/tmp/qt/moc_create_feature_dialog.o ../out/debug/tmp/qt/moc_draw_widget.o ../out/debug/tmp/qt/moc_editor_dialog.o ../out/debug/tmp/qt/moc_info_dialog.o ../out/debug/tmp/qt/moc_mainwindow.o ../out/debug/tmp/qt/moc_osm_auth_dialog.o ../out/debug/tmp/qt/moc_place_page_dialog.o ../out/debug/tmp/qt/moc_preferences_dialog.o ../out/debug/tmp/qt/moc_search_panel.o ../out/debug/tmp/qt/moc_update_dialog.o   -L../out/debug -lmap -ldrape_frontend -lrouting -lsearch -lstorage -lindexer -ldrape -lplatform -leditor -lgeometry -lcoding -lbase -lfreetype -lexpat -lfribidi -ltomcrypt -ljansson -lprotobuf -losrm -lstats_client -lminizip -lsuccinct -lpugixml -loauthcpp -lopening_hours -lz -L/home/y/Qt/5.5/gcc_64/lib -lQt5OpenGL -L/usr/lib64 -lQt5Widgets -lQt5Gui -lQt5Network -lQt5Core -lGL -lpthread 
../../out/debug/libindexer.a(editable_map_object.o): In function `osm::EditableMapObject::ValidatePostCode(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)':
/mnt/linux-secondary-hdd/y/coding/omim/indexer/editable_map_object.cpp:252: undefined reference to `search::v2::LooksLikePostcode(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool)'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Makefile:222: recipe for target '../../out/debug/indexer_tests' failed
```
* ???
* PROFIT

Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: And specification of search after indexer in the *.pro file doesn't help, it'll just delay the problem - search heavily depends on indexer and this dependency order is wrong - in all *.pro files search must go before indexer.

File: indexer/indexer_tests/editable_map_object_test.cpp [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+UNIT_TEST(EditableMapObject_SetWebsite)
+{
+  EditableMapObject emo;
+  emo.SetWebsite("https://some.thing.org");
+  TEST_EQUAL(emo.GetWebsite(), "https://some.thing.org", ());
+
+  emo.SetWebsite("some.thing.org");
+  TEST_EQUAL(emo.GetWebsite(), "http://some.thing.org", ());
+
+  emo.SetWebsite("some.thing.org");
```
Comment (Zverik) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Почему два одинаковых теста?

File: indexer/indexer_tests/editable_map_object_test.cpp [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+
+UNIT_TEST(EditableMapObject_ValidateFlats)
+{
+  TEST(EditableMapObject::ValidateFlats(""), ());
+  TEST(EditableMapObject::ValidateFlats("123"), ());
+  TEST(EditableMapObject::ValidateFlats("123a"), ());
+  TEST(EditableMapObject::ValidateFlats("a"), ());
+  TEST(EditableMapObject::ValidateFlats("123-456;a-e"), ());
+  TEST(EditableMapObject::ValidateFlats("123-456"), ());
+  TEST(EditableMapObject::ValidateFlats("123-456; 43-45"), ());
```
Comment (Zverik) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Добавь ещё такую же строку, но с запятой в качестве разделителя.

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 15 above)
Comment (Zverik) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Добавь сюда валидацию строк с `http://` и `https://` (и `ftp://`).

File: indexer/editable_map_object.cpp [hunk 20]
```diff
@@ -213,4 +191,139 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
 }
 
 void EditableMapObject::SetPointType() { m_geomType = feature::EGeomType::GEOM_POINT; }
+
+// static
+bool EditableMapObject::ValidateBuildingLevels(string const & buildingLevels)
+{
+  if (buildingLevels.size() > log10(kMaximumLevelsEditableByUsers) + 1)
```
Comment (Zverik) [inline indexer/editable_map_object.cpp]: Умно, но почему просто не оставить `2`?

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 19 above)
Comment (mgsergio) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Done.

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 18 above)
Comment (mgsergio) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 20 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Если решим разрешить проставлять этажи небоскребам, то поменять надо будет только в одном месте :)

**Comment by Zverik** [comment]
LGTM

**Comment by mgsergio** [comment]
@ygorshenin PTAL

File: indexer/editable_map_object.cpp (hunk 20 above)
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: Зачем так сложно. Проверь что buildingLevels.size() <= 18.

File: indexer/editable_map_object.cpp [hunk 21]
```diff
@@ -213,4 +191,139 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+  auto constexpr kMaxHouseNumberLength = 15;
+  if (us.size() > kMaxHouseNumberLength)
+    return false;
+
+  // TODO: Should we allow arabic numbers like U+0661 ١	Arabic-Indic Digit One?
+  strings::NormalizeDigits(us);
+  for (auto const c : us)
+  {
+    // Valid house numbers contain at least one digit.
+    if (c >= '0' && c <= '9')
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: IsASCIIDigit

File: indexer/editable_map_object.cpp (hunk 5 above)
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: Looks like we don't have IsLetter() in our API. Leave this as is.

File: indexer/editable_map_object.cpp [hunk 22]
```diff
@@ -213,4 +191,139 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+  return search::LooksLikePostcode(postCode, false /* check whole matching */);
+}
+
+// static
+bool EditableMapObject::ValidatePhone(string const & phone)
+{
+  if (phone.empty())
+    return true;
+
+  auto start = begin(phone);
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: s/start/curr/g

File: indexer/editable_map_object.cpp [hunk 23]
```diff
@@ -213,4 +191,139 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+}
+
+// static
+bool EditableMapObject::ValidateWebsite(string const & site)
+{
+  if (site.empty())
+    return true;
+
+  auto const dotPos = find(begin(site), end(site), '.');
+  // Site should contain at least one dot but not at the begining and not at the and.
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: but not at the beginning and the end.

File: indexer/indexer_tests/editable_map_object_test.cpp [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+
+  emo.SetWebsite("http://some.thing.org");
+  TEST_EQUAL(emo.GetWebsite(), "http://some.thing.org", ());
+
+  emo.SetWebsite("some.thing.org");
+  TEST_EQUAL(emo.GetWebsite(), "http://some.thing.org", ());
+
+  emo.SetWebsite("");
+  TEST_EQUAL(emo.GetWebsite(), "", ());
+
```
Comment (ygorshenin) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Please, remove this blank line.

File: indexer/indexer_tests/editable_map_object_test.cpp [hunk 25]
```diff
@@ -0,0 +1,114 @@
+#include "testing/testing.hpp"
+
+#include "indexer/editable_map_object.hpp"
+
+using osm::EditableMapObject;
+
+UNIT_TEST(EditableMapObject_SetWebsite)
```
Comment (ygorshenin) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Put to the anonymous namespace, please.

File: indexer/indexer_tests/editable_map_object_test.cpp [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+  TEST(!EditableMapObject::ValidateBuildingLevels("26"), ());
+  TEST(!EditableMapObject::ValidateBuildingLevels("ab"), ());
+  TEST(!EditableMapObject::ValidateBuildingLevels(
+      "2345534564564453645534545345534564564453645"), ());
+}
+
+UNIT_TEST(EditableMapObject_ValidateHouseNumber)
+{
+  TEST(EditableMapObject::ValidateHouseNumber(""), ());
+  TEST(EditableMapObject::ValidateHouseNumber("qwer7ty"), ());
```
Comment (ygorshenin) [inline indexer/indexer_tests/editable_map_object_test.cpp]: This validator looks insane. Have you ever meet such housenumbers in the OSM database?

File: indexer/indexer_tests/editable_map_object_test.cpp [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+  TEST(EditableMapObject::ValidatePhone("000 000 00"), ());
+  TEST(EditableMapObject::ValidatePhone("+00 0000 000 000"), ());
+
+  TEST(!EditableMapObject::ValidatePhone("+00 0000 000 0000 000"), ());
+  TEST(!EditableMapObject::ValidatePhone("00 00"), ());
+  TEST(!EditableMapObject::ValidatePhone("acb"), ());
+  TEST(!EditableMapObject::ValidatePhone("000 000 00b"), ());
+}
+
+// See validate phone.
```
Comment (ygorshenin) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Remove this comment, please.

Comment (ygorshenin) [inline indexer/indexer_tests/editable_map_object_test.cpp]: If ValidateFax is the same as ValidatePhone, why not to create a single function: ValidateFaxOrPhone?

File: indexer/postcodes_matcher.cpp [hunk 28]
```diff
@@ -107,15 +102,15 @@ class PostcodesMatcher
   // patterns.
   //
   // Complexity: O(total length of tokens in |slice|).
-  bool HasString(TokenSlice const & slice) const
+  bool HasString(StringSliceBase const & slice, bool handleAsPrefix) const
```
Comment (ygorshenin) [inline indexer/postcodes_matcher.cpp]: s/handleAsPrefix/isPrefix/g

File: indexer/postcodes_matcher.cpp [hunk 29]
```diff
@@ -168,8 +159,19 @@ PostcodesMatcher const & GetPostcodesMatcher()
 }
 }  // namespace
 
-bool LooksLikePostcode(TokenSlice const & slice) { return GetPostcodesMatcher().HasString(slice); }
+bool LooksLikePostcode(StringSliceBase const & slice, bool handleAsPrefix)
```
Comment (ygorshenin) [inline indexer/postcodes_matcher.cpp]: s/handleAsPrefix/isPrefix/g

File: indexer/postcodes_matcher.hpp [hunk 30]
```diff
@@ -0,0 +1,17 @@
...
+#include "indexer/string_slice.hpp"
+
+#include "std/cstdint.hpp"
+#include "std/string.hpp"
+
+namespace search
+{
+bool LooksLikePostcode(StringSliceBase const & slice, bool handleAsPrefix);
+/// Splits s into tokens and call LooksLikePostcode(TokenSlice) on the result.
+/// If handleAsPrefix is true returns true if some postcode starts with s.
```
Comment (ygorshenin) [inline indexer/postcodes_matcher.hpp]: Comment about handleAsPrefix should be near the first version of LooksLikePostcode.

File: search/v2/geocoder.cpp [hunk 31]
```diff
@@ -1021,7 +1021,7 @@ void Geocoder::WithPostcodes(TFn && fn)
         break;
 
       TokenSlice slice(m_params, startToken, startToken + n);
-      if (LooksLikePostcode(slice))
+      if (LooksLikePostcode(QuerySliceOnTokens(slice), true /*handleAsPrefix*/))
```
Comment (ygorshenin) [inline search/v2/geocoder.cpp]: This is not true. isPrefix should be true if and only if startToken + n == m_numTokens. Please, add a test on this case.

File: search/v2/token_slice.hpp [hunk 32]
```diff
@@ -67,36 +69,21 @@ class TokenSliceNoCategories
...
-
-  virtual TString const & Get(size_t i) const = 0;
-  virtual size_t Size() const = 0;
-  virtual bool IsPrefix(size_t i) const = 0;
-
-  bool Empty() const { return Size() == 0; }
-};
-
-class QuerySliceOnTokens : public QuerySlice
+class QuerySliceOnTokens : public StringSliceBase
```
Comment (ygorshenin) [inline search/v2/token_slice.hpp]: s/QuerySliceOnTokens/QuerySlice/g

File: indexer/editable_map_object.cpp (hunk 20 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Хватит занудствовать!

File: indexer/editable_map_object.cpp (hunk 23 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Протестую. Тогда вот так: beginning/end.

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 24 above)
Comment (mgsergio) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Done.

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 25 above)
Comment (mgsergio) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Done.

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 26 above)
Comment (mgsergio) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Done. Todo is set to editable_map_object.cpp:ValidateHouseNumber.

File: indexer/editable_map_object.cpp (hunk 21 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 22 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/indexer_tests/editable_map_object_test.cpp (hunk 27 above)
Comment (mgsergio) [inline indexer/indexer_tests/editable_map_object_test.cpp]: Done.

File: indexer/postcodes_matcher.cpp (hunk 28 above)
Comment (mgsergio) [inline indexer/postcodes_matcher.cpp]: Done.

File: indexer/postcodes_matcher.cpp (hunk 29 above)
Comment (mgsergio) [inline indexer/postcodes_matcher.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 23 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

File: indexer/postcodes_matcher.hpp (hunk 30 above)
Comment (mgsergio) [inline indexer/postcodes_matcher.hpp]: Done.

File: search/v2/token_slice.hpp (hunk 32 above)
Comment (mgsergio) [inline search/v2/token_slice.hpp]: Done.

File: search/v2/geocoder.cpp (hunk 31 above)
Comment (mgsergio) [inline search/v2/geocoder.cpp]: Done.

File: indexer/editable_map_object.cpp (hunk 20 above)
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: Не надо использовать числа с плавающей точкой там где их можно не использовать. Не говоря уж и о том что я не могу утверждать что с fast math все будет работать корректно.

Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: И нужна ли вообще эта провека? to_uint64 не справится?

File: indexer/editable_map_object.cpp [hunk 33]
```diff
@@ -213,4 +191,135 @@ void EditableMapObject::SetOpeningHours(string const & openingHours)
...
+        return false;
+    }
+  }
+  return true;
+}
+
+// static
+bool EditableMapObject::ValidatePostCode(string const & postCode)
+{
+  return search::LooksLikePostcode(postCode, false /* check whole matching */);
```
Comment (ygorshenin) [inline indexer/editable_map_object.cpp]: s/check whole matching/isPrefix/

File: indexer/string_slice.hpp [hunk 34]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include "base/string_utils.hpp"
+
+#include "std/string.hpp"
+#include "std/vector.hpp"
+
```
Comment (ygorshenin) [inline indexer/string_slice.hpp]: #include "std/cstdint.hpp"

Comment (mgsergio) [inline indexer/string_slice.hpp]: Done.

File: indexer/editable_map_object.cpp (hunk 33 above)
Comment (mgsergio) [inline indexer/editable_map_object.cpp]: Done.

**Comment by mgsergio** [comment]
@ygorshenin PTAL

**Comment by ygorshenin** [comment]
LGTM
